src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -155,60 +215,76 @@ internal void InitializeForSourceGen(JsonSerializerOptions options, JsonProperty
             JsonTypeInfo propertyTypeInfo = propertyInfo.PropertyTypeInfo;
             Type declaringType = propertyInfo.DeclaringType;
 
-            JsonConverter? converter = propertyInfo.Converter;
-            if (converter == null)
+            JsonConverter<T>? typedCustomConverter = propertyInfo.Converter;
+            CustomConverter = typedCustomConverter;
+
+            JsonConverter<T>? typedNonCustomConverter = propertyTypeInfo.Converter as JsonConverter<T>;
```

Comment:
Nit: rename "nonCustom" to something better like "typeLevelConverter" or "defaultConverterForType"